[build] Fork rules_multitool to our own extension.

aiuto opened the pull request:
Import the core parts of rules_multitool but with appropriate modifications for our needs.

- Re-root things so this is a local extension, rather than a distinct module.
- remove the ability to use .netrc for authentication.
- remove the :cwd and :workspace_root variations.
  - There are equivalent workarounds using `--run_under="cd <path> &&"`
  - Using `bazel run` is usually not the best practice. Rules should use `$(location <tool>)`. If a tool is so common that people need to call it anywhere, at any time, then it should be in their path.
  - The code is left commented out. Ready to be enabled if a good case is presented.
- Remove the support for WORKSPACE.

Next steps:
- Change download structure so the exeuctable name is the tool, rather than `"executable"`.
- Stop passing the attributes of each binary around as json blobs to be encoded and decoded.
- Make `:path` feature that will print the full execution path.

JSGette commented on bazel/multitool/extension.bzl:
```diff
@@ -0,0 +1,280 @@
+                file = archive_file,
+                url = binary["url"],
+            ))
+        rctx.symlink(archive_file, target_executable)
```
What does it do in case of Windows?

aiuto replied:
Great question. IIUC, the only practical way to use bazel on windows is to enable symlinks.

aiuto replied:
Either way.... symlinks is a general bazel issue, separate from this PR.

JSGette commented on bazel/multitool/private/lockfile.bzl:
```diff
@@ -0,0 +1,60 @@
+"Utilities for interacting with the multitool lockfile."
+
+def _check(condition, message):
+    "fails iff condition is False and emits message"
```
```suggestion
    "fails if condition is False and emits message"
```

aiuto replied:
done

JSGette commented on bazel/multitool/private/templates.bzl:
```diff
@@ -0,0 +1,105 @@
+        } | (substitutions or {}),
+    )
+
+def _renter_tool_labels(tools):
```
```suggestion
def _render_tool_labels(tools):
```

JSGette commented on bazel/multitool/private/templates.bzl:
```diff
@@ -0,0 +1,105 @@
+
+def _tools_subs(hub_name, tools):
+    return {
+        "{tool_labels}": _renter_tool_labels(tools),
```
```suggestion
        "{tool_labels}": _render_tool_labels(tools),
```

JSGette reviewed (approved):
All in all LGTM

ofek commented:
Out of curiosity, have we considered forking in a separate repository or do we have a requirement for committing everything related to Bazel in this repo?

aiuto committed:
[build] Fork rules_multitool to our own extension.

Import the core parts of rules_multitool but with appropriate modifications for our needs.

- Re-root things so this is a local extension, rather than a distinct module.
- remove the ability to use .netrc for authentication.
- remove the :cwd and :workspace_root variations.
  - There are equivalent workarounds using `--run_under="cd <path> &&"`
  - Using `bazel run` is usually not the best practice. Rules should use `$(location <tool>)`. If a tool is so common that people need to call it anywhere, at any time, then it should be in their path.
  - The code is left commented out. Ready to be enabled if a good case is presented.
- Remove the support for WORKSPACE.

Next steps:
- Change download structure so the exeuctable name is the tool, rather than `"executable"`.
- Stop passing the attributes of each binary around as json blobs to be encoded and decoded.
- Make `:path` feature that will print the full execution path.

aiuto committed:
[build] Fork rules_multitool to our own extension.

Import the core parts of rules_multitool but with appropriate modifications for our needs.

- Re-root things so this is a local extension, rather than a distinct module.
- remove the ability to use .netrc for authentication.
- remove the :cwd and :workspace_root variations.
  - There are equivalent workarounds using `--run_under="cd <path> &&"`
  - Using `bazel run` is usually not the best practice. Rules should use `$(location <tool>)`. If a tool is so common that people need to call it anywhere, at any time, then it should be in their path.
  - The code is left commented out. Ready to be enabled if a good case is presented.
- Remove the support for WORKSPACE.

Next steps:
- Change download structure so the exeuctable name is the tool, rather than `"executable"`.
- Stop passing the attributes of each binary around as json blobs to be encoded and decoded.
- Make `:path` feature that will print the full execution path.

[build] Fork rules_multitool to our own extension.

Import the core parts of rules_multitool but with appropriate modifications for our needs.

- Re-root things so this is a local extension, rather than a distinct module.
- remove the ability to use .netrc for authentication.
- remove the :cwd and :workspace_root variations.
  - There are equivalent workarounds using `--run_under="cd <path> &&"`
  - Using `bazel run` is usually not the best practice. Rules should use `$(location <tool>)`. If a tool is so common that people need to call it anywhere, at any time, then it should be in their path.
  - The code is left commented out. Ready to be enabled if a good case is presented.
- Remove the support for WORKSPACE.

Next steps:
- Change download structure so the exeuctable name is the tool, rather than `"executable"`.
- Stop passing the attributes of each binary around as json blobs to be encoded and decoded.
- Make `:path` feature that will print the full execution path.

Update bazel/multitool/extension.bzl

Yeah. That's a better name.

Co-authored-by: Joseph Gette <jgettepost@gmail.com>

Update bazel/multitool/extension.bzl

Co-authored-by: Joseph Gette <jgettepost@gmail.com>

Update bazel/multitool/private/templates.bzl

Co-authored-by: Joseph Gette <jgettepost@gmail.com>

render

iff

commit

lintyfresth

aiuto committed:
[build] Fork rules_multitool to our own extension.

Import the core parts of rules_multitool but with appropriate modifications for our needs.

- Re-root things so this is a local extension, rather than a distinct module.
- remove the ability to use .netrc for authentication.
- remove the :cwd and :workspace_root variations.
  - There are equivalent workarounds using `--run_under="cd <path> &&"`
  - Using `bazel run` is usually not the best practice. Rules should use `$(location <tool>)`. If a tool is so common that people need to call it anywhere, at any time, then it should be in their path.
  - The code is left commented out. Ready to be enabled if a good case is presented.
- Remove the support for WORKSPACE.

Next steps:
- Change download structure so the exeuctable name is the tool, rather than `"executable"`.
- Stop passing the attributes of each binary around as json blobs to be encoded and decoded.
- Make `:path` feature that will print the full execution path.
